Source/Core/InputCommon/ControllerInterface/SteamDeck/SteamDeck.cpp

```diff
@@ -0,0 +1,302 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "InputCommon/ControllerInterface/SteamDeck/SteamDeck.h"
+
+#include <type_traits>
+
+#include <hidapi.h>
+
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+#include "Common/MathUtil.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+namespace ciface::SteamDeck
+{
+constexpr std::string_view STEAMDECK_SOURCE_NAME = "SteamDeck";
+
+struct DeckInputReport
+{
+  // +0
+  u8 major_ver;
+  u8 minor_ver;
+  u8 report_type;
+  u8 report_sz;
+
+  u32 frame;
+
+  // +8
+  u32 buttons0;
+  u32 buttons1;
+
+  // +16
+  s16 l_pad_x;
+  s16 l_pad_y;
+  s16 r_pad_x;
+  s16 r_pad_y;
+
+  // +24
+  s16 accel_x;
+  s16 accel_y;
+  s16 accel_z;
+  s16 gyro_pitch;
+  s16 gyro_roll;
+  s16 gyro_yaw;
+  s16 pose_quat_w;
+  s16 pose_quat_x;
+  s16 pose_quat_y;
+  s16 pose_quat_z;
+
+  // +44
+  u16 l_trig;
+  u16 r_trig;
+
+  // +48
+  s16 l_stick_x;
+  s16 l_stick_y;
+  s16 r_stick_x;
+  s16 r_stick_y;
+
+  // +56
+  u16 l_pad_force;
+  u16 r_pad_force;
+  s16 l_stick_capa;
+  s16 r_stick_capa;
+};
+static_assert(sizeof(DeckInputReport) == 64);
+
+class Device final : public Core::Device
+{
+private:
+  class Button final : public Input
+  {
+  public:
+    Button(const char* name, const u32& buttons, u32 mask)
+        : m_name(name), m_buttons(buttons), m_mask(mask)
+    {
+    }
+    std::string GetName() const override { return m_name; }
+    ControlState GetState() const override { return (m_buttons & m_mask) != 0; }
+
+  private:
+    const char* const m_name;
+    const u32& m_buttons;
+    const u32 m_mask;
+  };
+
+  template <class T>
+  class AnalogInput : public Input
+  {
+  public:
+    AnalogInput(const char* name, const T& input, ControlState range)
+        : m_name(name), m_input(input), m_range(range)
+    {
+    }
+    std::string GetName() const final override { return m_name; }
+    ControlState GetState() const final override { return ControlState(m_input) / m_range; }
+
+  private:
+    const char* m_name;
+    const T& m_input;
+    const ControlState m_range;
+  };
+
+  class MotionInput final : public AnalogInput<s16>
+  {
+  public:
+    using AnalogInput::AnalogInput;
+    bool IsDetectable() const override { return false; }
+  };
+
+public:
+  Device(hid_device* device);
+  std::string GetName() const final override;
+  std::string GetSource() const final override;
+  void UpdateInput() override;
+
+private:
+  hid_device* m_device;
+  struct DeckInputReport m_latestInput;
```

Comment:
`DeckInputReport m_latest_input`

`struct` keyword is superfluous here and rename the variable to match our naming conventions.